### PR TITLE
Non-televised refactoring

### PR DESCRIPTION
**Pull request by filiph**
Some tidying up after today's boring show shoot. Introduces a lot of functionality that was supposed to be in the app but we never got to it.

hn_app_refactor

We're now also continually testing the app, so no more nasty surprises just before the shoot, hopefully.

This should put us in position to have an episode about something interesting (perf? error handling? testing? pagination?) without trying to dodge technical debt and unimplemented features.

### CODE REVIEW

File: lib/main.dart [hunk 1]
```diff
@@ -2,20 +2,30 @@ import 'package:flutter/foundation.dart';
...
 void main() {
   runApp(
     MultiProvider(
       providers: [
-        ValueListenableProvider(builder: (_) => ValueNotifier(true)),
-        ChangeNotifierProvider(builder: (_) => HackerNewsNotifier()),
-        Provider(builder: (_) => PrefsBloc()),
+        Provider<LoadingTabsCount>(
+          builder: (_) => LoadingTabsCount(),
+          dispose: (_, value) => value.dispose(),
```
Comment (efortuna) [inline lib/main.dart:19]: is this something you might want to talk about on-camera in an episode?

File: lib/main.dart [hunk 2]
```diff
@@ -48,75 +58,85 @@ class MyHomePage extends StatefulWidget {
...
+  final PageController _pageController = PageController();
+
+  @override
+  void initState() {
+    _pageController.addListener(_handlePageChange);
+    super.initState();
+  }
+
+  @override
+  void dispose() {
```
Comment (efortuna) [inline lib/main.dart:70]: same with this stuff

File: lib/main.dart [hunk 3]
```diff
@@ -48,75 +58,85 @@ class MyHomePage extends StatefulWidget {
...
 
   @override
   Widget build(BuildContext context) {
+    final hn = Provider.of<HackerNewsNotifier>(context);
+    final tabs = hn.tabs;
+    final current = tabs[_currentIndex];
+
+    if (current.articles.isEmpty && !current.isLoading) {
+      // New tab with no data. Let's fetch some.
+      current.refresh();
```
Comment (efortuna) [inline lib/main.dart:89]: should we do this in initState?

File: lib/main.dart [hunk 4]
```diff
@@ -48,75 +58,85 @@ class MyHomePage extends StatefulWidget {
...
         items: [
-          BottomNavigationBarItem(
-            title: Text('Top Stories'),
-            icon: Icon(Icons.arrow_drop_up),
-          ),
-          BottomNavigationBarItem(
-            title: Text('New Stories'),
-            icon: Icon(Icons.new_releases),
-          ),
+          for (final tab in tabs)
```
Comment (efortuna) [inline lib/main.dart:130]: I wonder if just:

`items: tabs.map((tab) => BottomNavigationBarItem(...)).toList()` is cleaner?

Just because this items list consists of exclusively the result of this for-loop. But I am certainly not the style person. Could be persuaded either way.

File: lib/src/notifiers/hn_api.dart [hunk 5]
```diff
@@ -2,58 +2,85 @@ import 'dart:async';
...
+  }
+
+  /// Articles from all tabs. De-duplicated.
+  UnmodifiableListView<Article> get allArticles => UnmodifiableListView(
+      _tabs.expand((tab) => tab.articles).toSet().toList(growable: false));
+
+  UnmodifiableListView<HackerNewsTab> get tabs => UnmodifiableListView(_tabs);
+}
+
+class HackerNewsTab with ChangeNotifier {
```
Comment (efortuna) [inline lib/src/notifiers/hn_api.dart:54]: recommend talking a bit in the next episode about the thinking behind this change

File: lib/src/widgets/search.dart [hunk 6]
```diff
@@ -32,73 +32,47 @@ class ArticleSearch extends SearchDelegate<Article> {
...
-      builder:
-          (context, AsyncSnapshot<UnmodifiableListView<Article>> snapshot) {
-        if (!snapshot.hasData) {
-          return Center(
-              child: Text(
-            'No data!',
-          ));
-        }
+    final results = articles
+        .where((a) => a.title.toLowerCase().contains(query.toLowerCase()));
```
Comment (efortuna) [inline lib/src/widgets/search.dart:61]: might talk about how the removal of bloc removes these StreamBuilder layers because ChangeNotifier now tells us when we need to redraw

File: lib/main.dart [hunk 7]
```diff
@@ -48,75 +58,85 @@ class MyHomePage extends StatefulWidget {
...
-      body: Consumer<HackerNewsNotifier>(
-        builder: (context, bloc, child) => ListView(
-              key: PageStorageKey(_currentIndex),
-              children: bloc.articles
-                  .map((a) => _Item(
-                        article: a,
-                        prefsBloc: Provider.of<PrefsBloc>(context),
-                      ))
-                  .toList(),
+      body: PageView.builder(
```
Comment (efortuna) [inline lib/main.dart:119]: talk about why this change on-camera

File: lib/main.dart [hunk 8]
```diff
@@ -48,75 +58,85 @@ class MyHomePage extends StatefulWidget {
...
         onTap: (index) {
-          if (index == 0) {
-            Provider.of<HackerNewsNotifier>(context)
-                .getStoriesByType(StoriesType.topStories);
-          } else {
-            assert(index == 1);
-            Provider.of<HackerNewsNotifier>(context)
-                .getStoriesByType(StoriesType.newStories);
-          }
+          _pageController.animateToPage(index,
```
Comment (efortuna) [inline lib/main.dart:137]: much cleaner. maybe highlight on-camera

File: lib/main.dart [hunk 9]
```diff
@@ -201,6 +216,41 @@ class _Item extends StatelessWidget {
   }
 }
 
+class _TabPage extends StatelessWidget {
```
Comment (efortuna) [inline lib/main.dart:219]: pulling this out is a good example of refactoring really simplifying code in a couple of places elsewhere. I'd recommend coding this up, or at least highlighting it in the next episode.

**Review by efortuna — APPROVED** [review]
LGTM! I have a number of places I recommend you talk about stuff in the Boring Show. If you want me to be more selective in my "recommendations" I can do a stricter pass, to limit my suggestions. :-)

**Comment by filiph** [comment]
Thank you Emily!

**Comment by dmiedev** [comment]
It's a pity that you haven't recorded this!

**Comment by chaadow** [comment]
The use of PageController and Nested providers when building items (and more) 💯 This Pull Request definitely deserves an episode of its own  👌 Great refactoring @filiph
